kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/RangeAwarePortPerNodeClusterNetworkAddressConfigProvider.java

```diff
@@ -0,0 +1,139 @@
+        this.exclusivePorts = Collections.unmodifiableSet(exclusivePorts);
+    }
+
+    @Override
+    public HostPort getClusterBootstrapAddress() {
+        return this.bootstrapAddress;
+    }
+
+    @Override
+    public HostPort getBrokerAddress(int nodeId) throws IllegalArgumentException {
+        if (!nodeIdIndex.contains(nodeId)) {
+            throw new IllegalArgumentException(
+                    "Cannot generate node address for node id %d as it is not contained in the ranges defined for provider with downstream bootstrap %s"
+                            .formatted(
+                                    nodeId,
+                                    bootstrapAddress));
+        }
+        int port = nodeStartPort + nodeIdIndex.indexOf(nodeId);
+        return new HostPort(BrokerAddressPatternUtils.replaceLiteralNodeId(nodeAddressPattern, nodeId), port);
+    }
+
+    @Override
+    public Set<Integer> getExclusivePorts() {
+        return this.exclusivePorts;
+    }
+
+    @Override
+    public Map<Integer, HostPort> discoveryAddressMap() {
+        return nodeIdIndex.values().stream()
+                .collect(Collectors.toMap(Function.identity(), this::getBrokerAddress));
+    }
+
+    /**
+     * @param name arbitrary name for this range, no functional purpose
+     * @param range interval spec
+     * @see IndexedIntSet#parseInterval(String)
+     */
+    public record Range(String name, String range) {};
+
+    /**
+     * Creates the configuration for this provider.
+     */
+    public static class RangeAwarePortPerNodeClusterNetworkAddressConfigProviderConfig {
+        private final HostPort bootstrapAddress;
+        private final String nodeAddressPattern;
+        private final int nodeStartPort;
+        private final IndexedIntSet nodeIdIndex;
+
+        public RangeAwarePortPerNodeClusterNetworkAddressConfigProviderConfig(@JsonProperty(required = true) HostPort bootstrapAddress,
+                                                                              @JsonProperty(required = false) String nodeAddressPattern,
+                                                                              @JsonProperty(required = false) Integer nodeStartPort,
+                                                                              @JsonProperty(required = true) List<Range> nodeIdRanges) {
+            Objects.requireNonNull(bootstrapAddress, "bootstrapAddress cannot be null");
+            if (nodeIdRanges.isEmpty()) {
+                throw new IllegalArgumentException("node id ranges empty");
+            }
+            this.bootstrapAddress = bootstrapAddress;
+            this.nodeAddressPattern = nodeAddressPattern != null ? nodeAddressPattern : bootstrapAddress.host();
+            this.nodeStartPort = nodeStartPort != null ? nodeStartPort : (bootstrapAddress.port() + 1);
+            Stream<IndexedIntSet> rangeSets = nodeIdRanges.stream().map(range -> IndexedIntSet.parseInterval(range.range));
```

Comment:
we not actually making any use of the Range#name.  My idea was to use it in error reporting (if ranges overlapped).  Maybe that idea is excessive.